src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventProvider.cs

```diff
@@ -1124,13 +741,76 @@ protected internal unsafe bool WriteEvent(ref EventDescriptor eventDescriptor, I
     // A wrapper around the ETW-specific API calls.
     internal sealed class EtwEventProvider : EventProviderImpl
     {
-        private readonly WeakReference<EventProvider> _eventProvider;
-        private long _registrationHandle;
-        private GCHandle _gcHandle;
+        /// <summary>
+        /// A struct characterizing ETW sessions (identified by the etwSessionId) as
+        /// activity-tracing-aware or legacy. A session that's activity-tracing-aware
+        /// has specified one non-zero bit in the reserved range 44-47 in the
+        /// 'allKeywords' value it passed in for a specific EventProvider.
+        /// </summary>
+        public struct SessionInfo
+        {
+            internal int sessionIdBit;      // the index of the bit used for tracing in the "reserved" field of AllKeywords
+            internal int etwSessionId;      // the machine-wide ETW session ID
+
+            internal SessionInfo(int sessionIdBit_, int etwSessionId_)
+            { sessionIdBit = sessionIdBit_; etwSessionId = etwSessionId_; }
+        }
+
+        private readonly WeakReference<EventProvider> m_eventProvider;
+        private long m_registrationHandle;
+        private GCHandle m_gcHandle;
+        private List<SessionInfo>? m_liveSessions;       // current live sessions (KeyValuePair<sessionIdBit, etwSessionId>)
+        private Guid m_providerId;
 
         internal EtwEventProvider(EventProvider eventProvider)
         {
-            _eventProvider = new WeakReference<EventProvider>(eventProvider);
+            m_eventProvider = new WeakReference<EventProvider>(eventProvider);
+        }
+
+        internal override void Disable()
+        {
+            base.Disable();
+            m_liveSessions = null;
+        }
+
+        protected override unsafe void HandleSessionEnable(
+                                    EventProvider target,
+                                    byte *additionalData,
+                                    byte level,
+                                    long matchAnyKeywords,
+                                    long matchAllKeywords,
+                                    Interop.Advapi32.EVENT_FILTER_DESCRIPTOR* filterData)
+        {
+            Debug.Assert(additionalData == null);
+
+            List<KeyValuePair<SessionInfo, bool>> sessionsChanged = GetSessions();
```

Comment:
The original code had a big comment suggesting that sessions didn't work. Do you feel confident it does now work or do we need to preserve parts of that comment?